src/tx_builder/builder.rs

```diff
@@ -0,0 +1,331 @@
+    unlock::ScriptUnlocker,
+    Address, HumanCapacity, NetworkInfo, ScriptId,
+};
+use ckb_jsonrpc_types as json_types;
+use ckb_types::core::TransactionView;
+use ckb_types::packed::{Byte32, OutPoint};
+use ckb_types::H256;
+use ckb_types::{
+    bytes::Bytes,
+    core::{BlockView, FeeRate},
+    packed::{CellDep, CellOutput, Script, WitnessArgs},
+    prelude::*,
+};
+
+use super::TxBuilderError;
+
+/// base transaction builder
+pub struct BaseTransactionBuilder {
+    pub ckb_client: CkbRpcClient,
+    pub sender: Address,
+    pub outputs: Vec<(CellOutput, Bytes)>,
+    pub cell_deps: HashSet<CellDep>,
+    pub network_info: NetworkInfo,
+
+    pub cell_collector: Box<dyn CellCollector>,
+    pub cell_dep_resolver: Box<dyn CellDepResolver>,
+    pub header_dep_resolver: Box<dyn HeaderDepResolver>,
+    pub tx_dep_provider: Box<dyn TransactionDependencyProvider>,
+    pub balancer: CapacityBalancer,
+    pub unlockers: HashMap<ScriptId, Box<dyn ScriptUnlocker>>,
+}
+
+pub trait CkbTransactionBuilder {
+    fn build_base(&mut self) -> Result<TransactionView, TxBuilderError>;
+    /// build a balanced transaction
+    fn build_balanced(&mut self) -> Result<TransactionView, TxBuilderError>;
+    /// build a balanced transaction and unlocked
+    fn build_unlocked(&mut self) -> Result<(TransactionView, Vec<ScriptGroup>), TxBuilderError>;
+    /// build a balanced transaction and check cycle limit and the transaction is unlocked.
+    fn build_balance_unlocked(
+        &mut self,
+    ) -> Result<(TransactionView, Vec<ScriptGroup>), TxBuilderError>;
+}
+
+impl BaseTransactionBuilder {
+    /// build a new BaseTransactionBuilder.
+    /// The placeholder_witness is default with 65 bytes lock, can be replaced later with set_placeholder_witness.
+    /// # Arguments:
+    ///   * `network_info` network type and url
+    ///   * `sender` the sender address string
+    pub fn new(
+        network_info: NetworkInfo,
+        sender_addr: &str,
+    ) -> Result<BaseTransactionBuilder, TxBuilderError> {
+        let sender_address =
+            Address::from_str(sender_addr).map_err(TxBuilderError::AddressFormat)?;
+
+        Self::new_with_address(network_info, sender_address)
+    }
+
+    pub fn new_with_address(
+        network_info: NetworkInfo,
+        sender_address: Address,
+    ) -> Result<BaseTransactionBuilder, TxBuilderError> {
+        let mut ckb_client = CkbRpcClient::new(network_info.url.as_str());
+        let cell_dep_resolver = {
+            let genesis_block = ckb_client.get_block_by_number(0.into())?.unwrap();
+            DefaultCellDepResolver::from_genesis(&BlockView::from(genesis_block)).unwrap()
+        };
+
+        let cell_collector = DefaultCellCollector::new(&network_info.url);
+        let tx_dep_provider = DefaultTransactionDependencyProvider::new(&network_info.url, 10);
+
+        let placeholder_witness = WitnessArgs::new_builder()
+            .lock(Some(Bytes::from(vec![0u8; 65])).pack())
```

Comment:
why hardcode 65 here? I think it's only for the built-in `secp256k1signhashall` lock script, if user pass other type `send_address` to this fn, it will build an invalid witness layout.